toolflow/vivado/arch/common/arch.tcl

```diff
@@ -35,40 +35,65 @@ namespace eval arch {
-          dict set ret $intf "interface $intf [format "offset 0x%08x range 0x%08x" $offset $range] kind register"
-          incr offset $range
+      set reg_segs [lsort [get_bd_addr_segs -filter { USAGE == register } $pe/*]]
+      set mem_segs [lsort [get_bd_addr_segs -filter { USAGE == memory } $pe/*]]
+      if {[llength $reg_segs] <= 1 && [llength $mem_segs] <= 1} {
+        puts "  processing $pe registers ..."
+        for {set i 0} {$i < [llength $reg_segs]} {incr i} {
+          set seg [lindex $reg_segs $i]
+          puts "    seg: $seg"
+          if {[get_property MODE [get_bd_intf_pins -of_objects $seg]] == "Master"} {
+            puts "    skipping master seg $seg"
+          } else {
+            set intf [get_bd_intf_pins -of_objects $seg]
+            set range [get_property RANGE $seg]
+            set offset [next_valid_address $offset $range]
+            ::platform::addressmap::add_processing_element [llength [dict keys $ret]] $offset $range
+            dict set ret $intf "interface $intf [format "offset 0x%08x range 0x%08x" $offset $range] kind register"
+            incr offset $range
+          }
         }
-      }
-      puts "  processing $pe memories ..."
-      set usrs [lsort [get_bd_addr_segs -filter { USAGE == memory } $pe/*]]
-      for {set i 0} {$i < [llength $usrs]} {incr i} {
-        set seg [lindex $usrs $i]
-        puts "    seg: $seg"
-        if {[get_property MODE [get_bd_intf_pins -of_objects $seg]] == "Master"} {
-          puts "    skipping master seg $seg"
-          continue
-        } else {
-          set intf [get_bd_intf_pins -of_objects $seg]
-          set range [get_property RANGE $seg]
-          set offset [next_valid_address $offset $range]
-          ::platform::addressmap::add_processing_element [llength [dict keys $ret]] $offset $range
-          dict set ret $intf "interface $intf [format "offset 0x%08x range 0x%08x" $offset $range] kind memory"
-          incr offset $range
+        puts "  processing $pe memories ..."
+        for {set i 0} {$i < [llength $mem_segs]} {incr i} {
+          set seg [lindex $mem_segs $i]
+          puts "    seg: $seg"
+          if {[get_property MODE [get_bd_intf_pins -of_objects $seg]] == "Master"} {
+            puts "    skipping master seg $seg"
+          } else {
+            set intf [get_bd_intf_pins -of_objects $seg]
+            set range [get_property RANGE $seg]
+            set offset [next_valid_address $offset $range]
+            ::platform::addressmap::add_processing_element [llength [dict keys $ret]] $offset $range
+            dict set ret $intf "interface $intf [format "offset 0x%08x range 0x%08x" $offset $range] kind memory"
+            incr offset $range
+          }
+        }
+      } else {
+        # if there is more than one reg/mem interface, we assume that the user knows what they are doing and add them in the same order
```

Comment:
"Knows what they are doing" in this context means that the user knows how to address the different interfaces? So this requires changes to the runtime, correct?

Reply:
When a PE uses local memory the tapasco runtime can only access it, if it is located in the address segment directly after the address segment of the control registers.
This does not work if the PE has multiple control registers or local memories.
Hence, if the user provides a PE with many such interfaces, it is best to add them in the order provided by the user.

Reply:
This change was made in order to avoid changes to the runtime

Reply:
Yeah, I assumed something similar. I think the best case would be, if we could adapt the runtime accordingly so that this case is automatically detected and handled correctly. However I'm not sure how big of a change this would be and thus if it is feasible.

Reply:
if we had a PE with 4 interfaces:
ctrl-intf-A + mem-intf-A, ctrl-intf-B + mem-intf-B
Then the old tapasco would arrange them as
ctrl-intf-A, ctrl-intf-B, mem-intf-A, mem-intf-B
The runtime could detect this automatically by associating the first control regs with the first mem, and so on

However, if the PE has these interfaces
ctrl-intf-A + mem-intf-A, ctrl-intf-B, ctrl-intf-C + mem-intf-C
Then I don't think the runtime can associate control and memory correctly without additional user input

Reply:
It should be possible to include this information in the status core so that the runtime can automatically detect it. However, again not sure how feasible this is

Reply:
If I understand correctly, this would be part of #181, which requires similar changes to the status core.

Reply:
added a clarifying comment